pkg/clusteragent/externalmetrics/metrics_retriever.go

```diff
@@ -70,15 +70,21 @@ func (mr *MetricsRetriever) retrieveMetricsValues() {
 		return
 	}
 
-	queries, queriesMaxAge := getUniqueQueries(datadogMetrics)
-	log.Debugf("Starting refreshing external metrics with: %d queries", len(queries))
-
-	results, err := mr.processor.QueryExternalMetric(queries, queriesMaxAge)
+	queriesByTimeWindow := getUniqueQueriesByTimeWindow(datadogMetrics)
+	resultsByTimeWindow := make(map[time.Duration]map[string]autoscalers.Point)
 	globalError := false
-	// Check for global failure
-	if len(results) == 0 && err != nil {
-		globalError = true
-		log.Errorf("Unable to fetch external metrics: %v", err)
+
+	for timeWindow, queries := range queriesByTimeWindow {
+		log.Debugf("Starting refreshing external metrics with: %d queries (window: %d)", len(queries), timeWindow)
+
+		results, err := mr.processor.QueryExternalMetric(queries, timeWindow)
```

Comment:
I don't think the logic is entirely accurate, because a DatadogMetric with `TimeWindow || MaxAge < defaultWindow` will be put in the `map` with this value with the query will be executed with `defaultWindow` due to the defaulting in `QueryExternalMetric`.

There are currently 4 parameters:
```
external_metrics_provider.bucket_size = default 5m
external_metrics_provider.max_age = default 2m

metric.TimeWindow = default not set
metric.MaxAge = default not set
```

We have:
```
defaultWindow = max(external_metrics_provider.bucket_size, external_metrics_provider.max_age)
We have a single time window with min(max(defaultWindow, all(metric.MaxAge)), 24h)
```

I think that `defaultWindow` is still the same. When computing query `timeWindow` the formula should be `max(defaultWindow, metric.MaxAge, metric.TimeWindow)` to still batch all queries up to `defaultWindow`.

This should be done here instead of `QueryExternalMetric` and the other callers of `QueryExternalMetric` should be modified to pass a valid `timeWindow`. Only the limit check should be done in `QueryExternalMetric`.

Reply:
I think this is done